pumpkin-protocol/src/packet_encoder.rs

```diff
@@ -49,7 +49,7 @@ impl PacketEncoder {
 
                 self.compress_buf.clear();
 
-                let data_len_size = VarInt(data_len as i32).written_size();
+                let data_len_size = VarInt::try_from(data_len).unwrap().written_size();
```

Comment:
You added a bunch of unwraps in packet_encoder. This looks pretty dangerous. The first rule in Networking is to never trust the Client.